cpp/leaky_core.cc

```diff
@@ -181,10 +184,31 @@ bool InputAvailable(int fd) {
   NotImplemented();
 }
 
-void SignalState_AfterForkingChild() {
-  signal(SIGQUIT, SIG_DFL);
-  signal(SIGPIPE, SIG_DFL);
-  signal(SIGTSTP, SIG_DFL);
+static void signal_handler(int sig_num) {
+  assert(gSignalHandlers != nullptr);
+  SignalHandler* handler = gSignalHandlers->get(sig_num);
```

Comment:
There is likely a race here, because the main thread is writing and signal handlers are reading from this Dict

I think the "bottom-up" strategy of doing SharedTrapState and TSAN will help, and this code may end up looking different


This is simlar to the `std::map` demo in `cpp/data_race_test.cc` -- whether it's a race depends on the internal implementation of the `Dict`, and it almost certainly is (we could test it with TSAN!)

Also our dict implementation is likely to change (it doesn't have hashing yet), and it's definitely not intended to be used unsynchronized from different threads of control

Reply:
Oh wow this is a pretty glaring hole in this...

Reply:
I _think_ I've convinced myself we don't need this Dict here. Pushed a fixup commit that we can talk about when we chat next. I think the challenge will be making the usage from python more obviously correct